### ISSUE TITLE
Rethink sanitizer architecture

### ISSUE THREAD

**Issue by svenjacobs** [svenjacobs/leon#50]
Currently sanitizers are purely based on regular expressions and can only remove parts from a URL. However feature requests have shown that this concept is not flexible enough since sanitizers could also for example [extract](https://github.com/svenjacobs/leon/issues/42#issuecomment-1141794303) parts from, [rewrite](https://github.com/svenjacobs/leon/issues/45) or [decode](https://github.com/svenjacobs/leon/issues/49) URLs.

The current sanitizers are stored in a SQLite database on the device. The idea was to have some kind of form in a future version of Leon where users can add their own sanitizers / rules. However this concept does not work with more flexible sanitizers. Prospectively sanitizers will not be stored in a database. But sanitizers could provide their own UI. For instance there could be a `CustomParametersSanitizer` which provides a UI for specifying own URL parameters to be removed. How these parameters are stored is implementation detail of the sanitizer. It could be shared preferences or SQLite.

The question remains what should be shown on the result screen where currently the number of removed parameters are shown? Since future sanitizers must not necessarily remove parameters, different metrics must be displayed?

**Comment by TPS** [comment]
> The question remains what should be shown on the result screen where currently the number of removed parameters are shown? Since future sanitizers must not necessarily remove parameters, different metrics must be displayed?

Could this be programmable as part of the provided UI?  Also, these metrics are nice to have, but not actually important, so couldn't they be optional?  I'd think the only required parts for the programmable UI would be orginal & the post-processed URLs.  Anything else could be left off.

**Comment by svenjacobs** [comment]
> Could this be programmable as part of the provided UI?  Also, these metrics are nice to have, but not actually important, so couldn't they be optional?  I'd think the only required parts for the programmable UI would be orginal & the post-processed URLs.  Anything else could be left off.

I guess I will remove these metrics for now and only show the original and modified URL. The UI of a sanitizer was meant to be placed on the settings screen.

**Comment by svenjacobs** [comment]
Work on this has been completed. This should unblock further issues / feature requests.

**Comment by TPS** [comment]
Maybe instead show before/after URL?  That'd probably be easier than calculating these metrics, anyway.
